### ISSUE TITLE
Inconsistent placement rules when retrying truncating tables/partitions

### ISSUE THREAD

**Issue by xhebox** [pingcap/tidb#31540]
## Bug Report

https://github.com/pingcap/tidb/blob/4f30a14256c1d51adcd4fdcef53ba837fa0c8438/ddl/table.go#L603-L611

https://github.com/pingcap/tidb/blob/aa7ad03bcd4ae85ef3a31cfccf2c908df9521d6f/ddl/partition.go#L1177-L1192

Since truncating tables/partitions will allocate id just before placement rules operations, every retrying job will lead to placement rules of different IDs. It breaks idempotence.

It is reported by @CalvinNeo when testing tiflash placement rules, caused by `writing conflict of concurrent DDL`.

While it is a correctness problem, retrying only occurs with heavy DDL load. As an experimental feature, it does not need to be critical or major bug.

### 1. Minimal reproduce step (Required)

Start many concurrent session to execute truncate DDLs.

### 2. What did you expect to see? (Required)

Writing conflict leads to retrying, but retrying jobs have same IDs.

### 3. What did you see instead (Required)

Writing conflict leads to retrying, but retrying jobs have different IDs.

**Comment by xhebox** [comment]
There are two possible solutions in my mind.:

1. we move id allocation into the first submit phase of DDL job
2. middle state truncation job, split id allocation and placement rules into two phases.

I prefer the solution 1, simple and easy, compatibility is possible with a careful taking on job arguments. Solution 2 is good for compatibility since no need of job argument changes. But it does need a new middle state, which is different from previous DDLs.

**Comment by lcwangchao** [comment]
> There are two possible solutions in my mind.:
> 
> 1. we move id allocation into the first submit phase of DDL job
> 2. middle state truncation job, split id allocation and placement rules into two phases.
> 
> I prefer the solution 1, simple and easy, compatibility is possible with a careful taking on job arguments. Solution 2 is good for compatibility since no need of job argument changes. But it does need a new middle state, which is different from previous DDLs.

For solution 1, we still need to handle some corner cases. Because the ddl operation is concurrent before inqueue, it can happens that after allocating the new partition ids, the number of partition changes. Though it may be not so easy to happen but I think we should add some checks...

**Comment by CalvinNeo** [comment]
If there are eventually redundant PD rules for TiFlash A, and if PD can't schedule regions away from TiFlash A, it may cause failure when scale in.

**Comment by xhebox** [comment]
> 



> > There are two possible solutions in my mind.:
> > 
> > 1. we move id allocation into the first submit phase of DDL job
> > 2. middle state truncation job, split id allocation and placement rules into two phases.
> > 
> > I prefer the solution 1, simple and easy, compatibility is possible with a careful taking on job arguments. Solution 2 is good for compatibility since no need of job argument changes. But it does need a new middle state, which is different from previous DDLs.
> 
> For solution 1, we still need to handle some corner cases. Because the ddl operation is concurrent before inqueue, it can happens that after allocating the new partition ids, the number of partition changes. Though it may be not so easy to happen but I think we should add some checks...

From the code, partitions to be truncated are decided at the time of submiting/enqueuing. `pids` are responsible for that:

https://github.com/pingcap/tidb/blob/aa7ad03bcd4ae85ef3a31cfccf2c908df9521d6f/ddl/ddl_api.go#L3480

So we don't actually truncate new partitions if new partitions are added after submitting. I don't think there is a problem, only if see the current behavior is a bug..

**Comment by xhebox** [comment]
> If there are eventually redundant PD rules for TiFlash A, and if PD can't schedule regions away from TiFlash A, it may cause failure when scale in.

It does not affect the behavior, at least for the current code base. As long as the id is allocated increasing monotonically. But it does damage the performance of PD, if there are hundreds of redundant PD rules.

**Comment by CalvinNeo** [comment]
> > If there are eventually redundant PD rules for TiFlash A, and if PD can't schedule regions away from TiFlash A, it may cause failure when scale in.
> 
> It does not affect the behavior, at least for the current code base. As long as the id is allocated increasing monotonically. But it does damage the performance of PD, if there are hundreds of redundant PD rules.

In current version of TiFlash Cluster Manager(the old one that is in use), all pd rules are set when TableID is allocated, so will result in no redundant pd rules.

**Comment by xhebox** [comment]
> I think these two behaviors are different

Yeah, but you misinterpret me. I mean redundant rules does not affect the schedule of PD, since id is allocated increasing monotonically, which means redundant rules are just no-op rules.

**Comment by CalvinNeo** [comment]
> > I think these two behaviors are different
> 
> Yeah, but you misinterpret me. I mean redundant rules does not affect the schedule of PD, since id is allocated increasing monotonically, which means redundant rules are just no-op rules.

So, the redundant rule `table-%v-r` set for TiFlash actually manages a non-existing table. So there will be no scheduled region from PD at the beginning?

**Comment by xhebox** [comment]
> > > I think these two behaviors are different
> > 
> > 
> > Yeah, but you misinterpret me. I mean redundant rules does not affect the schedule of PD, since id is allocated increasing monotonically, which means redundant rules are just no-op rules.
> 
> So, the redundant rule `table-%v-r` set for TiFlash actually manages a non-existing table. So there will be no scheduled region from PD at the beginning?

Yes, rules are applied to regions. If there are no regions(no finished DDL to generate regions of that ID), rules are, well, redundant, as it says. And it does not affect the schedule at all.

**Comment by xhebox** [comment]
Should have been fixed.
